File: Classes/FuelStatisticsPageController.m
// FuelStatisticsPageController.m
//
// Kraftstoff


#import "FuelStatisticsPageController.h"
#import "FuelStatisticsGraphViewController.h"
#import "FuelStatisticsTextViewController.h"
#import "AppDelegate.h"



@implementation FuelStatisticsPageController
{
    BOOL pageControlUsed;
}



#pragma mark -
#pragma mark View Lifecycle



- (instancetype)initWithNibName:(NSString *)nibName bundle:(NSBundle *)nibBundle
{
    if ((self = [super initWithNibName:nibName bundle:nibBundle])) {

        self.modalTransitionStyle = UIModalTransitionStyleCrossDissolve;
    }

    return self;
}


- (void)viewDidLoad
{
    [super viewDidLoad];

    // Load content pages
    for (NSInteger page = 0; page < _pageControl.numberOfPages; page++) {

        FuelStatisticsViewController *controller = nil;

        switch (page) {

            case 0:controller = [FuelStatisticsViewController_PriceDistance  alloc]; break;
            case 1:controller = [FuelStatisticsViewController_AvgConsumption alloc]; break;
            case 2:controller = [FuelStatisticsViewController_PriceAmount alloc]; break;
            case 3:controller = [FuelStatisticsTextViewController alloc]; break;
        }

        controller = [controller initWithNibName:@"FuelStatisticsViewController" bundle:nil];
        controller.selectedCar = self.selectedCar;

        [self addChildViewController:controller];
        controller.view.frame = [self frameForPage:page];

        [_scrollView addSubview:controller.view];
    }

    // Configure scroll view
    _scrollView.contentSize = CGSizeMake (StatisticsViewWidth * _pageControl.numberOfPages, StatisticsViewHeight);
    _scrollView.scrollsToTop = NO;

    // Enlarge scrollView, hide pageControl
    _scrollView.frame = self.view.frame;
    _pageControl.hidden = YES;

    // Select preferred page
    dispatch_async (dispatch_get_main_queue(), ^{

        _pageControl.currentPage = [[NSUserDefaults standardUserDefaults] integerForKey:@"preferredStatisticsPage"];
        [self scrollToPage:_pageControl.currentPage animated:NO];

        pageControlUsed = NO;
    });
    
    [[NSNotificationCenter defaultCenter]
        addObserver:self
           selector:@selector(localeChanged:)
               name:NSCurrentLocaleDidChangeNotification
             object:nil];

    [[NSNotificationCenter defaultCenter]
        addObserver:self
           selector:@selector(didEnterBackground:)
               name:UIApplicationDidEnterBackgroundNotification
             object:nil];

    [[NSNotificationCenter defaultCenter]
        addObserver:self
           selector:@selector(didBecomeActive:)
               name:UIApplicationDidBecomeActiveNotification
             object:nil];

    [[NSNotificationCenter defaultCenter]
        addObserver:self
           selector:@selector(numberOfMonthsSelected:)
               name:@"numberOfMonthsSelected"
             object:nil];
}


- (UIStatusBarStyle)preferredStatusBarStyle
{
    return UIStatusBarStyleLightContent;
}


- (void)viewWillDisappear:(BOOL)animated
{
    [super viewWillDisappear:animated];

    [[NSUserDefaults standardUserDefaults] setInteger:_pageControl.currentPage forKey:@"preferredStatisticsPage"];
    [[NSUserDefaults standardUserDefaults] synchronize];
}


#pragma mark -
#pragma mark View Rotation



- (BOOL)shouldAutorotate
{
    return YES;
}


- (NSUInteger)supportedInterfaceOrientations
{
    return UIInterfaceOrientationMaskLandscape;
}



#pragma mark -
#pragma mark Cache Handling



- (void)invalidateCaches
{
    for (FuelStatisticsViewController *controller in self.childViewControllers)
        [controller invalidateCaches];
}



#pragma mark -
#pragma mark System Events



- (void)localeChanged:(id)object
{
    [self invalidateCaches];
}


- (void)didEnterBackground:(id)object
{
    [[NSUserDefaults standardUserDefaults] setInteger:_pageControl.currentPage forKey:@"preferredStatisticsPage"];
    [[NSUserDefaults standardUserDefaults] synchronize];

    for (FuelStatisticsViewController *controller in self.childViewControllers)
        [controller purgeDiscardableCacheContent];
}


- (void)didBecomeActive:(id)object
{
    [self updatePageVisibility];
}



#pragma mark -
#pragma mark User Events



- (void)numberOfMonthsSelected:(NSNotification*)notification
{
    // Remeber selection in preferences
    NSInteger numberOfMonths = [[[notification userInfo] valueForKey:@"span"] integerValue];
    [[NSUserDefaults standardUserDefaults] setInteger:numberOfMonths forKey:@"statisticTimeSpan"];

    // Update all statistics controllers
    for (NSInteger i = 0; i < _pageControl.numberOfPages; i++) {

        NSInteger page = (_pageControl.currentPage + i) % _pageControl.numberOfPages;

        FuelStatisticsViewController *controller = (self.childViewControllers)[page];
        [controller setDisplayedNumberOfMonths:numberOfMonths];
    }
}



#pragma mark -
#pragma mark Frame Computation for Pages



- (CGRect)frameForPage:(NSInteger)page
{
    page = [_scrollView visiblePageForPage:page];

    CGRect frame = _scrollView.frame;
    frame.origin.x = frame.size.width * page;
    frame.origin.y = 0;

    return frame;
}



#pragma mark -
#pragma mark Sync ScrollView with Page Indicator



- (void)scrollViewDidScroll:(UIScrollView*)sender
{
    if (pageControlUsed == NO) {

        NSInteger newPage = floor ((_scrollView.contentOffset.x - StatisticsViewWidth*0.5) / StatisticsViewWidth) + 1;
        newPage = [self.scrollView pageForVisiblePage:newPage];

        if (_pageControl.currentPage != newPage) {

            _pageControl.currentPage = newPage;
            [self updatePageVisibility];
        }
    }
}


- (void)scrollViewWillBeginDragging:(UIScrollView*)scrollView
{
    pageControlUsed = NO;
}


- (void)scrollViewDidEndDecelerating:(UIScrollView*)view
{
    pageControlUsed = NO;
}



#pragma mark -
#pragma mark Page Control Handling



- (void)updatePageVisibility
{
    for (NSInteger page = 0; page < _pageControl.numberOfPages; page++) {

        FuelStatisticsViewController *controller = (self.childViewControllers)[page];
        [controller noteStatisticsPageBecomesVisible:(page == _pageControl.currentPage)];
    }
}


- (void)scrollToPage:(NSInteger)page animated:(BOOL)animated;
{
    pageControlUsed = YES;

    [_scrollView scrollRectToVisible:[self frameForPage:page] animated:animated];
    [self updatePageVisibility];
}


- (IBAction)pageAction:(id)sender
{
    [self scrollToPage:_pageControl.currentPage animated:YES];
}



#pragma mark -
#pragma mark Memory Management


- (void)dealloc
{
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

@end

File: Classes/FuelStatisticsPageController.swift
//
//  FuelStatisticsPageController.swift
//  kraftstoff
//
//  Created by Ingmar Stein on 05.05.15.
//
//

import UIKit

final class FuelStatisticsPageController: UIPageViewController {
  // Set by presenting view controller
  var selectedCar: Car!
  var statisticsViewControllers = [FuelStatisticsViewController]()
  let priceDistanceDataSource = FuelStatisticsViewControllerDataSourcePriceDistance()
  let avgConsumptionDataSource = FuelStatisticsViewControllerDataSourceAvgConsumption()
  let priceAmountDataSource = FuelStatisticsViewControllerDataSourcePriceAmount()

  // MARK: - View Lifecycle

  required init?(coder aDecoder: NSCoder) {
    super.init(coder: aDecoder)

    modalTransitionStyle = .crossDissolve
    dataSource = self
    delegate = self
  }

  override func viewDidLoad() {
    super.viewDidLoad()

    // Load content pages
    // swiftlint:disable:next force_cast
    let priceDistanceViewController = storyboard!.instantiateViewController(withIdentifier: "FuelStatisticsGraphViewController") as! FuelStatisticsGraphViewController
    priceDistanceViewController.dataSource = priceDistanceDataSource
    priceDistanceViewController.selectedCar = selectedCar
    priceDistanceViewController.pageIndex = 0

    // swiftlint:disable:next force_cast
    let avgConsumptionViewController = storyboard!.instantiateViewController(withIdentifier: "FuelStatisticsGraphViewController") as! FuelStatisticsGraphViewController
    avgConsumptionViewController.dataSource = avgConsumptionDataSource
    avgConsumptionViewController.selectedCar = selectedCar
    avgConsumptionViewController.pageIndex = 1

    // swiftlint:disable:next force_cast
    let priceAmountViewController = storyboard!.instantiateViewController(withIdentifier: "FuelStatisticsGraphViewController") as! FuelStatisticsGraphViewController
    priceAmountViewController.dataSource = priceAmountDataSource
    priceAmountViewController.selectedCar = selectedCar
    priceAmountViewController.pageIndex = 2

    // swiftlint:disable:next force_cast
    let statisticsViewController = storyboard!.instantiateViewController(withIdentifier: "FuelStatisticsTextViewController") as! FuelStatisticsTextViewController
    statisticsViewController.selectedCar = selectedCar
    statisticsViewController.pageIndex = 3

    statisticsViewControllers = [priceDistanceViewController, avgConsumptionViewController, priceAmountViewController, statisticsViewController]

    DispatchQueue.main.async {
      var page = UserDefaults.standard.integer(forKey: "preferredStatisticsPage")
      if page < 0 || page >= self.statisticsViewControllers.count {
        page = 0
      }
      self.setViewControllers([self.statisticsViewControllers[page]], direction: .forward, animated: false, completion: nil)
    }

    NotificationCenter.default.addObserver(self,
                                           selector: #selector(FuelStatisticsPageController.localeChanged(_:)),
                                           name: NSLocale.currentLocaleDidChangeNotification,
                                           object: nil)

    NotificationCenter.default.addObserver(self,
                                           selector: #selector(FuelStatisticsPageController.didEnterBackground(_:)),
                                           name: UIApplication.didEnterBackgroundNotification,
                                           object: nil)

    NotificationCenter.default.addObserver(self,
                                           selector: #selector(FuelStatisticsPageController.didBecomeActive(_:)),
                                           name: UIApplication.didBecomeActiveNotification,
                                           object: nil)

    NotificationCenter.default.addObserver(self,
                                           selector: #selector(FuelStatisticsPageController.numberOfMonthsSelected(_:)),
                                           name: NSNotification.Name(rawValue: "numberOfMonthsSelected"),
                                           object: nil)
  }

  var currentPage: Int {
    guard let statisticsViewController = viewControllers?.first as? FuelStatisticsViewController else { return -1 }
    return statisticsViewController.pageIndex
  }

  override func viewWillDisappear(_ animated: Bool) {
    super.viewWillDisappear(animated)

    UserDefaults.standard.set(currentPage, forKey: "preferredStatisticsPage")
    UserDefaults.standard.synchronize()
  }

  // MARK: - View Rotation

  override var supportedInterfaceOrientations: UIInterfaceOrientationMask {
    .landscape
  }

  override var prefersStatusBarHidden: Bool {
    true
  }

  // MARK: - Cache Handling

  func invalidateCaches() {
    for controller in children {
      if let fuelStatisticsViewController = controller as? FuelStatisticsViewController {
        fuelStatisticsViewController.invalidateCaches()
      }
    }
  }

  // MARK: - System Events

  @objc func localeChanged(_: AnyObject) {
    invalidateCaches()
  }

  @objc func didEnterBackground(_: AnyObject) {
    UserDefaults.standard.set(currentPage, forKey: "preferredStatisticsPage")
    UserDefaults.standard.synchronize()

    for controller in statisticsViewControllers {
      controller.purgeDiscardableCacheContent()
    }
  }

  fileprivate func updatePageVisibility() {
    for controller in statisticsViewControllers {
      if viewControllers!.contains(controller) {
        controller.noteStatisticsPageBecomesVisible()
      }
    }
  }

  @objc func didBecomeActive(_: AnyObject) {
    updatePageVisibility()
  }

  // MARK: - User Events

  @objc func numberOfMonthsSelected(_ notification: Notification) {
    // Remember selection in preferences
    if let numberOfMonths = (notification as NSNotification).userInfo?["span"] as? Int {
      UserDefaults.standard.set(numberOfMonths, forKey: "statisticTimeSpan")

      // Update all statistics controllers
      for controller in children {
        if let fuelStatisticsViewController = controller as? FuelStatisticsViewController {
          fuelStatisticsViewController.displayedNumberOfMonths = numberOfMonths
        }
      }
    }
  }
}

extension FuelStatisticsPageController: UIPageViewControllerDelegate {
  func pageViewController(_: UIPageViewController, didFinishAnimating _: Bool, previousViewControllers _: [UIViewController], transitionCompleted completed: Bool) {
    if completed {
      updatePageVisibility()
    }
  }
}

extension FuelStatisticsPageController: UIPageViewControllerDataSource {
  func pageViewController(_: UIPageViewController, viewControllerAfter viewController: UIViewController) -> UIViewController? {
    guard let statisticsViewController = viewController as? FuelStatisticsViewController else { return nil }
    let page = statisticsViewController.pageIndex
    if statisticsViewController.pageIndex < statisticsViewControllers.count - 1 {
      return statisticsViewControllers[page + 1]
    } else {
      return statisticsViewControllers.first!
    }
  }

  func pageViewController(_: UIPageViewController, viewControllerBefore viewController: UIViewController) -> UIViewController? {
    guard let statisticsViewController = viewController as? FuelStatisticsViewController else { return nil }
    let page = statisticsViewController.pageIndex
    if page > 0 {
      return statisticsViewControllers[page - 1]
    } else {
      return statisticsViewControllers.last!
    }
  }

  func presentationCount(for _: UIPageViewController) -> Int {
    statisticsViewControllers.count
  }

  func presentationIndex(for _: UIPageViewController) -> Int {
    currentPage
  }
}
